### ISSUE TITLE
"helm upgrade --reuse-values --set=..." does not merge correctly

### ISSUE THREAD

**Issue by michaelkunzmann-sap** [helm/helm#4792]
Output of `helm version`:
```
Client: &version.Version{SemVer:"v2.9.1", GitCommit:"20adb27c7c5868466912eebdf6664e7390ebe710", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.9.1", GitCommit:"20adb27c7c5868466912eebdf6664e7390ebe710", GitTreeState:"clean"}
```

Output of `kubectl version`:
```
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.6", GitCommit:"9f8ebd171479bec0ada837d7ee641dec2f8c6dd1", GitTreeState:"clean", BuildDate:"2018-03-21T15:21:50Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"9+", GitVersion:"v1.9.7-gke.6", GitCommit:"9b635efce81582e1da13b35a7aa539c0ccb32987", GitTreeState:"clean", BuildDate:"2018-08-16T21:33:47Z", GoVersion:"go1.9.3b4", Compiler:"gc", Platform:"linux/amd64"}
```

Cloud Provider/Platform (AKS, GKE, Minikube etc.): GKE

Let's assume I have the following values:
```
$ helm get values myrelease
a:
  b:
    c: "foobar"
  c: "foobar"
b: "foobar"
```
When I run this:
```
helm upgrade myrelease my/chart --reuse-values --recreate-pods --set a.b.c=somethingelse
```

The entry ```a.b.c.``` will be updated, but ```a.c``` will be discarded:
```
$ helm get values myrelease
a:
  b:
    c: "somethingelse"
  // c: "foobar" is gone
b: "foobar"
```

**Comment by AdrienKuhn** [comment]
I have the exact same issue but with `helm template`

```
$ helm version
Client: &version.Version{SemVer:"v2.11.0", GitCommit:"2e55dbe1fdb5fdb96b75ff144a339489417b146b", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.11.0", GitCommit:"2e55dbe1fdb5fdb96b75ff144a339489417b146b", GitTreeState:"clean"}
```

**Comment by JustDoItSascha** [comment]
Is this bug really fixed? Because I also have it in version 3.6.2...? Also with helm template, too...
